### DISCUSSION TITLE
Render components inline / statically

### DISCUSSION THREAD

**Discussion by trueChazza** [livewire/livewire#5380] [3. Feature ideas]
👋 Hi, thank you for Livewire!

Would having the ability (and ease) to statically render inline and return the entire markup of a component, including the layout, be a useful addition to Livewire core?

#### Maybe? Could you explain

Currently in tests we have the ability to assert that we see / or don't see a given string in the HTML.
If I wanted to assert various states a component can have, I can very easily accomplish this with `assertSeeHtml` or `assertDontSeeHtml` - which is great.

However as a component begins to grow, naturally your tests grow with it - asserting its various states. Blindspots begin to become apparent when factoring up different states for different test cases. 

If 1 test case assumes a component will have a certain state, from another test case - how can I confidently see this (as a user would) without manually loading up the entire project and verifying?

#### Ok, show me some examples

Testing a component is as easy as:

```php
Livewire::actingAs(User::factory()->createQuietly())
    ->test(SomeComponent::class)
    ->assertSeeHtml('User Message');
```

When I need to test another state:

```php
Livewire::actingAs(User::factory()->createQuietly())
    ->test(SomeComponent::class)
    ->assertSeeHtml('User Message');


Livewire::actingAs(AnotherUser::factory()->createQuietly())
    ->test(SomeComponent::class)
    ->assertSeeHtml('Another User Message');
```

This only shows 2 examples - but imagine 200 cases with varying conditional states. How can I (and others) be confident that each state is rendered as expected (again as a user would see it), without loading up the entire project to see?

#### Right, so what are you proposing?

```php
TestableLivewire::macro('render', function () {
    $render = view($this->lastRenderedView->getData()['livewireLayout']['view'], [
        'slot' => $this->lastRenderedView
    ])
        ->render();

    return htmlspecialchars_decode($render);
});

Livewire::actingAs(User::factory()->createQuietly())
    ->test(SomeComponent::class)
    ->assertSeeHtml('User Message')
    ->render();
});
```

Adding a simple macro like this returns a rendered view the browser can display. This could be returned in a route for example to easily see the rendered state of the component. Or could even be exported to a static HTML file via Laravel's `Storage` facade to be used further.

I haven't dug too far into how the API could potentially look from a Livewire core perspective, but I'm opening this discussion first to see if this would be something of interest for me to continue further.

**Comment by joshhanley** [comment]
@trueChazza sounds like it could be a good idea, not sure if I would use it myself. At the moment though, I would probably just keep using it as a macro in your own project. The reason being is we are trying to limit new features going into V2, so it minimises how much needs to be rewritten into V3. Once V3 is out, feel free to raise the idea again or submit a PR if you think it would be useful 🙂

**Reply by trueChazza** [reply]
Sounds good, thanks for responding 😄

**Reply by joshhanley** [reply]
No worries! 🙂
